Various fixes + Pocket Loader & Placer Rewrite

CreepyCre opened the pull request:
### Current fixes:
- Rift targets now serialize correctly
- position printed out when teleporting via door now works correctly


### Loader & Placer rewrite
The goal of the loader & placer rewrite is to allow for more pockets than just schematic based pockets.
It is already completely functional, current features are:
- loads pockets of different types from json
- can randomize pocket selection with dynamic weights
- use pockets conditionally based on depth


Pocket placement uses VirtualPockets. There are 2 general VirtualPocket types:
1. pocket generators:
Generate pockets, self explanatory,
2. pocket selection modifier
Contain pocket generators and modifiy how these pocket generators are selected/ weighted. May be nested to modify pocket selection multiple times.


#### Current generator types are:
- SchematicGenerator:
Represents a pocket loaded from a schematic.
- ChunkGenerator (~70% done)
Generates a pocket based on the ChunkGenerator of a world specified. Size and offset may be configured.

#### Current selection modifiers are:
- DepthDependentSelector:
Contains a List of String - VirtualPocket pairs, where the String is used as a regex to compare against depth. Will place first VirtualPocket whose regex matches the current depth.

CreepyCre committed:
initial rewrite of pocket placement
instead of each pocket having an assigned template, the pockets now generate/ place themselves
allows for pockets with custom generators instead of schematics

CreepyCre committed:
Merge remote-tracking branch 'origin/1.16-fabric' into 1.16-fabric

CreepyCre committed:
rewrote weighted pocket system. Pockets now get supplied additional information to calculate their weight

CreepyCre committed:
added depth dependent pockets
somewhat flattened the pocket json

CreepyCre committed:
fixed depth?

BoogieMonster1O1 commented on src/main/java/org/dimdev/dimdoors/pockets/VirtualPocket.java:
```diff
@@ -0,0 +1,58 @@
+
+
+
+		Codec<T> getCodec();
```
I’d rather manually deserialize here.
They’re quite unforgiving and don’t handle exceptions well.
The whole reason I couldn’t get dungeons to work for so long was because of bad exception handling.
Accepting a Function<T, CompoundTag> and Function<CompoundTag, T> and calling them works well.

CreepyCre replied:
resolved https://github.com/DimensionalDevelopment/DimDoors/pull/184/commits/0fb93068bd17ffbd1cf192b22a015593fe6472cc

BoogieMonster1O1 commented on src/main/java/org/dimdev/dimdoors/pockets/VirtualDepthDependentPocket.java:
```diff
@@ -0,0 +1,82 @@
+
+
+	private final String name;
+	private final List<Pair<String, VirtualPocket>> pocketList;
```
why not just a Map<String, VirtualPocket>

CreepyCre replied:
it needs to retain the same order, not sure if a map would guarantee that

BoogieMonster1O1 replied:
LinkedHashMap?

BoogieMonster1O1 commented on src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java:
```diff
@@ -31,10 +27,9 @@
-    private static final Random RANDOM = new Random(new Random().nextLong());
+    private final Map<Identifier, PocketTemplateV2> templates = Maps.newHashMap();
+    private final Map<String, WeightedList<VirtualPocket, PocketGenerationParameters>> templateMap = Maps.newHashMap(); //TODO: un-ugly-fy
+    private final List<PocketGroup> pocketTypes = Lists.newArrayList();
```
pocketGroups

CreepyCre replied:
do you mean renaming the fields? How about:
``templateMap`` -> ``weightedPocketGroups``
``pocketTypes`` -> ``pocketGroups``

BoogieMonster1O1 replied:
even better 😅

BoogieMonster1O1 reviewed (commented):
Looks good so far!

Waterpicker commented:
Gonna admit. I dropped a brick when I saw this pop up.

CreepyCre committed:
some cleaning and renaming
generators generate pockets directly
selectors/ selection modifiers modify which generator gets selected or how it gets selected

CreepyCre committed:
renamed fields for clarity

CreepyCre committed:
fixed rift serialization

i509VCB commented on src/main/java/org/dimdev/dimdoors/pockets/SchematicV2Handler.java:
```diff
@@ -58,9 +53,17 @@ public void load() {
+
+                this.pocketGroups.add(type);
+
+				WeightedList<VirtualPocket, PocketGenerationParameters> weightedPockets = new WeightedList<>();
```
Indentation mismatch here

CreepyCre replied:
Doesn't appear that way on my end, but I noticed that everything seems to have been indented using four spaces previously, while I indented everything using tabs. Not sure how to ensure correct whitespace type.

BoogieMonster1O1 replied:
Enable editorconfig in your ide and that should hopefully make everything use tabs

CreepyCre committed:
switched VirtualPockets from Codecs to fromTag/ toTag serialization

CreepyCre committed:
fix overwriting VirtualPockets with same key

CreepyCre committed:
added ChunkGenerator pocket generator type, ~70% done

CreepyCre committed:
ChunkGenerator now generates Structures

CreepyCre committed:
fix ChunkGenerator + slight improvements
